interop-testing/src/test/java/io/grpc/testing/integration/TrafficControlProxy.java

```diff
@@ -0,0 +1,281 @@
+
+  /**
+   * Convenience method to change bandwidth/latency without restarting.
+   *
+   * @param targetBPS
+   * @param targetLatency
+   * @param latencyUnits
+   * @throws IOException
+   */
+  public void reset(int targetBPS, int targetLatency, TimeUnit latencyUnits) throws IOException {
+    checkArgument(targetBPS > 0);
+    checkArgument(targetLatency > 0);
+    shutDown();
+    bandwidth = targetBPS;
+    latency = latencyUnits.toNanos(targetLatency) / 2;
+    queueLength = (int) Math.max(bandwidth * latency / TimeUnit.SECONDS.toNanos(1), 1);
+    chunkSize = Math.max(1, queueLength);
+    serverQueueSize = 0;
+    clientQueueSize = 0;
+    serverQueue.clear();
+    clientQueue.clear();
+    workers.clear();
+    shutDown = false;
+    start();
+  }
+
+  /**
+   * Waits for client and server and start reader/writer threads.
+   *
+   * @throws IOException
+   */
+  public void start() throws IOException {
+    // ClientAcceptor uses a ServerSocket server so that the client can connect to the proxy as it
+    // normally would a server. serverSock then connects the server using a regular Socket as a
+    // client normally would.
+    logger.info("Starting new proxy on port " + proxyPort + " with Queue Length " + queueLength);
+    clientAcceptor = new ServerSocket();
+    clientAcceptor.setReuseAddress(true);
+    clientAcceptor.bind(new InetSocketAddress(loopback, proxyPort));
+    clientSock = clientAcceptor.accept();
+    serverSock = new Socket();
+    serverSock.connect(new InetSocketAddress(loopback, serverPort));
+
+    System.out.println("client and server connected");
+
+    startWorkers();
+    System.out.println("workers started");
+  }
+
+  public void shutDown() throws IOException {
+    // TODO: Handle case where a socket fails to close, therefore blocking the others from closing
+    logger.info("Proxy shutting down... ");
+    shutDown = true;
+    clientAcceptor.close();
+    clientSock.close();
+    serverSock.close();
+    for (Thread worker : workers) {
+      worker.interrupt();
+    }
+    logger.info("Shutdown Complete");
+  }
+
+  private void startWorkers() throws IOException {
+    DataInputStream clientIn = new DataInputStream(clientSock.getInputStream());
+    DataOutputStream clientOut = new DataOutputStream(serverSock.getOutputStream());
+    DataInputStream serverIn = new DataInputStream(serverSock.getInputStream());
+    DataOutputStream serverOut = new DataOutputStream(clientSock.getOutputStream());
+
+    workers.add(new Thread(new Reader(Endpoint.CLIENT, clientIn, clientQueue)));
+    workers.add(new Thread(new Writer(Endpoint.CLIENT, clientOut, clientQueue)));
+    workers.add(new Thread(new Reader(Endpoint.SERVER, serverIn, serverQueue)));
+    workers.add(new Thread(new Writer(Endpoint.SERVER, serverOut, serverQueue)));
+
+    for (Thread worker : workers) {
+      worker.start();
+    }
+
+  }
+
+  private void incrementQueue(Endpoint endpoint, int delta) {
+    if (endpoint == Endpoint.SERVER) {
```

Comment:
Put all methods that interact with the DelayQueue in a new class MessageQueue, but I left the Reader and Writer runnables as their own classes that take a MessageQueue as an argument.